[RFC] twister: implement `simulation_exclude` filter

ycsin opened the pull request:
This PR attempts to solve the dilemma when a physical board can pass a testcase but its simulator couldn't.

See:

https://github.com/zephyrproject-rtos/zephyr/blob/bd227e19cff9855355468974f1f984ed88f2cf3d/tests/kernel/timer/timer_behavior/testcase.yaml#L9-L19

and https://github.com/zephyrproject-rtos/zephyr/pull/66369#issuecomment-1849703114

The issue that I'm trying to solve in this PR is related to `renode`, but this patch should apply to other mcu-simulator as well

ycsin committed:
twister: fix a minor typo

Should be 'related' instead of 'realted'.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin committed:
twister: implement `simulation_exclude` filter

This gives us the flexibility to exclude certain platforms'
simulator that's known to fail from running in Twister, but
allows real hardware to be tested.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin committed:
tests: kernel: timer_behavior: exclude renode instead of boards

We just need to exclude renode simulation and not the physical
boards, let's do just that.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

PerMac commented:
I see the yaml you're editing contains:
```
   platform_type:
      - mcu	 
``` 
So it seems the test is already designed and marked as to be "mcu" specific. Maybe you could improve the handling of this type in twister, instead of adding new marking?

gchwier commented on scripts/pylib/twister/twisterlib/testplan.py:
```diff
@@ -756,6 +759,11 @@ def apply_filters(self, **kwargs):
                     if ts.platform_exclude and plat.name in ts.platform_exclude:
                         instance.add_filter("In test case platform exclude", Filters.TESTSUITE)
 
+                    if not device_testing \
```
why `device_testing` is checked here?

ycsin replied:
Because if Twister is running on a real board, the result shouldn't be affected by its `simulation` type

gchwier replied:
ok, I see. Platforms like `mps2_an385` have QEMU enabled.

ycsin replied:
> Platforms like `mps2_an385` have QEMU enabled.

Improvements like having the ability to use `qemu` instead of `renode` would be nice in this case, would have to change the schema and make additional changes to the Twister for that

ycsin commented:
> Maybe you could improve the handling of this type in twister, instead of adding new marking

Real boards and renode-simulated boards share the same `platform_type: mcu`, we will need something else to differentiate between the two

PerMac commented on scripts/pylib/twister/twisterlib/testplan.py:
```diff
@@ -756,6 +759,11 @@ def apply_filters(self, **kwargs):
                     if ts.platform_exclude and plat.name in ts.platform_exclude:
                         instance.add_filter("In test case platform exclude", Filters.TESTSUITE)
 
+                    if not device_testing \
+                            and ts.simulation_exclude \
+                            and plat.simulation in ts.simulation_exclude:
+                        instance.add_filter("In test case simulation exclude", Filters.SIMULATION)
```
My point is, that maybe you can check if the satisfying result would be obtained without introduction of new yaml entry (simulation_exclude) but with using the already existing one from test yamls: platform_type: mcu. Something like

```suggestion
                    if not device_testing and ts.platform_type == ['mcu']:
                        instance.add_filter("Test is made solely for on-target execution", Filters.SIMULATION)
```

ycsin replied:
Wouldn't this filter out `ts.platform_type == ['mcu']` as long as not testing on physical boards?

PerMac replied:
I believe this is the point of `platform_type` with only `mcu` inside, stating that the test is meant only for mcu. My concern is that with filtering I proposed such tests won't also build without "device-testing". But on the other hand just building, for such tests doesn't bring much value IMO. BTW it seems there are only 9 cases, where "platform_type" is used in test yaml, and this particular (kernel.timer) is the only one which has only 'mcu' listed.

ycsin replied:
> I believe this is the point of `platform_type` with only `mcu` inside

I think I get what you mean now..

ycsin replied:
I [tried this direction today](https://github.com/zephyrproject-rtos/zephyr/compare/main...ycsin:pr/twister_simulation_exclude1?expand=1) but it's confusing to assume that `type: mcu` means 'real hardware', given that it might have `qemu` or `renode` as its underlying simulator. Forcing tests to run only on real hardware when `platform_type` == `['mcu']` would affect other simulator(s) that could have passed the test.

hakehuang commented on scripts/schemas/twister/testsuite-schema.yaml:
```diff
@@ -156,6 +156,9 @@ mapping:
         sequence:
           - type: str
+      "simulation_exclude":
+        type: any
```
I would suggest to use seq instead of any, any is becuase we want backward compatible with string and list.

ycsin replied:
Thanks for the suggestion, I've updated it to use seq + enum

ycsin committed:
twister: change schema of `platform_type` to follow `type`

The `platform_type` was created to filter `type` so they
should have the same schema.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin commented:
I've moved the simulation filtering to `check_runnable` as it seems more appropriate, another plus is that tests will be built for platforms with blocked simulator. I also noticed that the test is running and passing on both `mps2_an385` & `mps2_an521_ns` even though they were in the `platform_exclude` list, and used `qemu` instead of `renode` as simulation

fkokosinski reviewed (dismissed):
LGTM

ycsin commented on tests/kernel/timer/timer_behavior/testcase.yaml:
```diff
@@ -6,17 +6,8 @@ tests:
-      - mpfs_icicle
-      - hifive_unleashed
-      - mps2_an385
-      - mps2_an521_ns
```
mps2 boards use qemu, and they seem to pass the test fine

ycsin commented on scripts/utils/twister_to_list.py:
```diff
@@ -25,6 +25,7 @@
     "extra_sections",
     "platform_exclude",
     "platform_allow",
+    "simulation_exclude",
```
fixed!
